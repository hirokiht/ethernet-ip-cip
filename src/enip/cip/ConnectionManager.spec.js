const { services, ForwardOpen, connection, transport, LargeForwardOpen, ForwardClose, UnconnectedSend, paths } = require('./ConnectionManager')

describe('Connection Manager', () => {
  describe('Constant Values', () => {
    it('services definition', () => {
      expect(services.FORWARD_CLOSE).toEqual(0X4E)
      expect(services.UNCONNECTED_SEND).toEqual(0X52)
      expect(services.FORWARD_OPEN).toEqual(0X54)
      expect(services.GET_CONNECTION_DATA).toEqual(0X56)
      expect(services.SEARCH_CONNECTION_DATA).toEqual(0X57)
      expect(services.GET_CONNECTION_OWNER).toEqual(0X5A)
      expect(services.LARGE_FORWARD_OPEN).toEqual(0X5B)
    })

    it('connection parameter', () => {
      expect(connection.REDUNDANT|connection.type.NULL|connection.priority.LOW|connection.size.FIXED).toEqual(1<<15)
      expect(connection.type.MULTICAST|connection.priority.HIGH|connection.size.VARIABLE).toEqual(1<<13|1<<10|1<<9)
      expect(connection.type.P2P|connection.priority.SCHEDULED|connection.size.FIXED).toEqual(2<<13|2<<10)
      expect(connection.type.RSVD|connection.priority.URGENT|connection.size.VARIABLE).toEqual(3<<13|3<<10|1<<9)
    })

    it('transport definition', () => {
      expect(transport.direction.CLIENT|transport.trigger.cyclic|transport.class[0]).toEqual(0)
      expect(transport.direction.CLIENT|transport.trigger.CHG_OF_STATE|transport.class[1]).toEqual(1<<4|1)
      expect(transport.direction.CLIENT|transport.trigger.APP_OBJ|transport.class[2]).toEqual(2<<4|2)
      expect(transport.direction.SERVER|transport.class[3]).toEqual(1<<7|3)
    })

    it('paths definition', () => {
      expect(paths.CONNECTION_MANAGER.toString('hex')).toEqual('20062401')
      expect(paths.MESSAGE_ROUTER.toString('hex')).toEqual('20022401')
    })
  })

  describe('ForwardOpen', () => {
    let asmPath = Buffer.from('2004246630643065')
    let otCon = connection.size.FIXED | connection.priority.LOW | connection.type.P2P
    let toCon = connection.size.VARIABLE | connection.priority.HIGH | connection.type.MULTICAST
    expect(ForwardOpen.build(100000, otCon, 255, toCon, 500, 0x1234, transport.direction.CLIENT|transport.class[1], paths.MESSAGE_ROUTER, 1024))
      .toMatchSnapshot()
    otCon = connection.size.VARIABLE | connection.priority.SCHEDULED | connection.type.MUTLICAST
    toCon = connection.size.FIXED | connection.priority.URGENT | connection.type.P2P
    expect(ForwardOpen.build(200000, toCon, 511, toCon, 200, 0x4321, transport.direction.SERVER|transport.class[3], asmPath, 2048))
      .toMatchSnapshot()
  })
})
