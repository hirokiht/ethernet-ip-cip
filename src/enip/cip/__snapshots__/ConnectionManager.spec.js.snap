// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Object {
  "data": Array [
    84,
    2,
    32,
    6,
    36,
    1,
    3,
    128,
    0,
    0,
    0,
    0,
    239,
    190,
    173,
    222,
    52,
    18,
    55,
    19,
    239,
    190,
    173,
    222,
    0,
    0,
    0,
    0,
    64,
    66,
    15,
    0,
    255,
    64,
    160,
    134,
    1,
    0,
    0,
    38,
    1,
    2,
    32,
    2,
    36,
    1,
  ],
  "type": "Buffer",
}
`;

exports[` 2`] = `
Object {
  "data": Array [
    84,
    2,
    32,
    6,
    36,
    1,
    4,
    128,
    0,
    0,
    0,
    0,
    239,
    190,
    173,
    222,
    33,
    67,
    55,
    19,
    239,
    190,
    173,
    222,
    0,
    0,
    0,
    0,
    64,
    66,
    15,
    0,
    255,
    77,
    64,
    13,
    3,
    0,
    0,
    76,
    131,
    8,
    50,
    48,
    48,
    52,
    50,
    52,
    54,
    54,
    51,
    48,
    54,
    52,
    51,
    48,
    54,
    53,
  ],
  "type": "Buffer",
}
`;

exports[` 3`] = `
Object {
  "data": Array [
    91,
    2,
    32,
    6,
    36,
    1,
    3,
    128,
    0,
    0,
    0,
    0,
    239,
    190,
    173,
    222,
    52,
    18,
    55,
    19,
    239,
    190,
    173,
    222,
    0,
    0,
    0,
    0,
    64,
    66,
    15,
    0,
    43,
    2,
    0,
    64,
    160,
    134,
    1,
    0,
    244,
    1,
    0,
    38,
    0,
    2,
    32,
    2,
    36,
    1,
  ],
  "type": "Buffer",
}
`;

exports[` 4`] = `
Object {
  "data": Array [
    91,
    2,
    32,
    6,
    36,
    1,
    4,
    128,
    0,
    0,
    0,
    0,
    239,
    190,
    173,
    222,
    33,
    67,
    55,
    19,
    239,
    190,
    173,
    222,
    0,
    0,
    0,
    0,
    64,
    66,
    15,
    0,
    255,
    1,
    0,
    76,
    64,
    13,
    3,
    0,
    176,
    4,
    0,
    76,
    130,
    8,
    50,
    48,
    48,
    52,
    50,
    52,
    54,
    54,
    51,
    48,
    54,
    52,
    51,
    48,
    54,
    53,
  ],
  "type": "Buffer",
}
`;

exports[` 5`] = `
Object {
  "data": Array [
    78,
    2,
    32,
    6,
    36,
    1,
    2,
    250,
    55,
    19,
    55,
    19,
    239,
    190,
    173,
    222,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[` 6`] = `
Object {
  "data": Array [
    78,
    2,
    32,
    6,
    36,
    1,
    2,
    250,
    55,
    19,
    55,
    19,
    239,
    190,
    173,
    222,
    8,
    0,
    50,
    48,
    48,
    52,
    50,
    52,
    54,
    54,
    51,
    48,
    54,
    52,
    51,
    48,
    54,
    53,
  ],
  "type": "Buffer",
}
`;

exports[` 7`] = `
Object {
  "data": Array [
    78,
    2,
    32,
    6,
    36,
    1,
    2,
    250,
    33,
    67,
    55,
    19,
    239,
    190,
    173,
    222,
    8,
    0,
    50,
    48,
    48,
    52,
    50,
    52,
    54,
    54,
    51,
    48,
    54,
    52,
    51,
    48,
    54,
    53,
  ],
  "type": "Buffer",
}
`;

exports[` 8`] = `
Object {
  "data": Array [
    82,
    2,
    32,
    6,
    36,
    1,
    3,
    250,
    24,
    0,
    48,
    101,
    48,
    53,
    50,
    49,
    48,
    48,
    48,
    49,
    48,
    48,
    50,
    53,
    48,
    48,
    48,
    50,
    48,
    48,
    51,
    48,
    48,
    51,
    2,
    0,
    48,
    49,
    48,
    49,
  ],
  "type": "Buffer",
}
`;
